### ISSUE TITLE
CURL error 6 since latest release

### ISSUE THREAD

**Issue by brazenvoid** [guzzle/guzzle#1532]
Our production site broke today as we deployed it with the guzzle updating to 6.2.1. Please tell me what changed? We have since locked the version down for guzzle to 6.2.0 for the foreseeable future on all our projects.

We are using Guzzle with Laravel 5.2.

**Comment by mtdowling** [comment]
The changelog and git history shows the changes. You probably also picked up an updated version of the PSR7 package: https://github.com/guzzle/psr7/blob/master/CHANGELOG.md#131---2016-06-25

What URL are you connecting to?

**Comment by chrisvensko** [comment]
We just ran into a similar issue. Looks like our URL objects were being cast to a string with 6.2.0, but were throwing an InvalidArgumentException with 6.2.1. Quick fix was to call $GuzzleClient->get((string) $URLObject, $options);

**Comment by chrisvensko** [comment]
Looking through the source, looks like this commit is where the breaking change happened: https://github.com/guzzle/guzzle/commit/443a7a621c82b619aa2c30c7119a7e8145b457fa

Before that, it would have called new Psr7\Uri($uri), which uses parse_url, which will take objects castable to a string.

**Comment by b8ne** [comment]
Im getting the same issue, tried casting to string but thats not working.
Just trying to process a simple ip get
`$client = new Client();
 $request = $client->get("http://ip-api.com/json/$ip");`
If there is no easy fix how best could I rollback this update?

**Comment by brazenvoid** [comment]
@b8ne setting the composer entry to 6.2.0 and running composer will downgrade it.

**Comment by brazenvoid** [comment]
@mtdowling we are calling our own API on a different server. Many sites use it. 

The problem is strange as if I pass the full URL of the API as base_uri in the client, it works but then it sets the request headers containing host names with domain prefixed with http/https. Certain other packages then fail in the API.

**Comment by b8ne** [comment]
@brazenvoid thanks mate.
Just found that casting did in fact work, it was however another request i was calling to "api.openweathermap.org"... I had to add http:// to the url for it to work.  Not sure of the reasoning but hopefully this helps somehow.

**Comment by chrisvensko** [comment]
Here is a code sample that works with 6.2.0 and fails with 6.2.1

Example code (successful at 6.2.0, fails at 6.2.1):
 
```php
class Url
{
    public function __toString() { return 'http://ip-api.com/json/127.0.0.1'; }
}

$client = new \GuzzleHttp\Client();
$response = $client->get(new Url(), []);

echo $response->getBody();
```

**Comment by taz77** [comment]
Make note of the reason for 6.2.1 for those that are considering rolling back:
https://github.com/guzzle/guzzle/blob/master/CHANGELOG.md

If you roll back you should implement some mitigation to prevent against the HTTP_PROXY vulnerability at the server or firewall level. This is explained in gross detail at https://httpoxy.org/

**Comment by mtdowling** [comment]
@Tobion can you take a look at these issues?

**Comment by Tobion** [comment]
> "api.openweathermap.org"... I had to add http:// to the url for it to work

@b8ne That problem should already be fixed by #1509, isn't it?

The other reports are about custom URI objects. But that has never been officially supported and was a misuse. See https://github.com/guzzle/guzzle/blob/master/src/ClientInterface.php#L48
The URI must be a string|UriInterface. Using anything else is your own risk.
Also the `base_uri` already validated before that it must be string|UriInterface. It was just missing for the `$uri` parameter. This shows that a custom Uri class was not meant to be supported.

**Comment by chrisvensko** [comment]
@Tobion, regardless of whether it was meant to be supported, it worked in versions prior to 6.2.1.

**Comment by Tobion** [comment]
@chrisvensko sure but that doesn't mean anything. If the contract is not followed, the result is nondeterministic. Otherwise we wouldn't need to document any parameters and it wouldn't be possible to change anything as every change might break anything for anybody who uses random parameter values or types.

**Comment by chrisvensko** [comment]
Sure, but these custom objects can be cast to a string.

**Comment by Tobion** [comment]
string type does not include objects with __toString

**Comment by brazenvoid** [comment]
@Tobion Can you give me some advice as to where to look in the code base to mitigate/investigate this? 

Simply previously we gave the URL without http/https prefixed in the request methods but now we have to prefix it and define it in base_uri in the client initialization. This results in the headers also having the http/https prefixed with the domain which in turn makes the Symphony getHost() method fail.

http://api.symfony.com/2.8/Symfony/Component/HttpFoundation/Request.html#method_getHost

Thanks.

**Comment by Tobion** [comment]
> headers also having the http/https prefixed

Sounds like the URI parsing failed to distinguish the scheme and the host. You could check what your URI object looks like at this point: https://github.com/guzzle/guzzle/blob/master/src/Client.php#L148

**Comment by Tobion** [comment]
Can be closed since #1538 is merged
